privacyidea/lib/tokens/tantoken.py

```diff
@@ -0,0 +1,190 @@
+#
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+This file contains the definition of the paper token class
+It depends on the DB model, and the lib.tokenclass.
+"""
+
+import logging
+from privacyidea.lib.log import log_with
+from privacyidea.lib.tokenclass import TokenClass
+from privacyidea.lib.tokens.papertoken import PaperTokenClass
+from privacyidea.lib.policy import SCOPE
+from privacyidea.lib import _
+from privacyidea.lib.policydecorators import libpolicy
+from privacyidea.lib.crypto import geturandom, hash
+import binascii
+
+log = logging.getLogger(__name__)
+DEFAULT_COUNT = 100
+SALT_LENGTH = 4
+
+
+class TANACTION(object):
+    TANTOKEN_COUNT = "papertoken_count"
+
+
+class TanTokenClass(PaperTokenClass):
+    """
+    The TAN token allows to print out the next e.g. 100 OTP values.
+    This sheet of paper can be used to authenticate and strike out the used
+    OTP values. In contrast to the paper token, the OTP values of the TAN token
+    can be used randomly.
+    """
+
+    @log_with(log)
+    def __init__(self, db_token):
+        """
+        This creates a new TAN token object from a DB token object.
+
+        :param db_token: instance of the orm db object
+        :type db_token:  orm object
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"tan")
+        self.hKeyRequired = False
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the token type shortname
+
+        :return: 'paper'
+        :rtype: string
+        """
+        return "tan"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        Return the prefix, that is used as a prefix for the serial numbers.
+        :return: PITN
+        """
+        return "PITN"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        returns a subtree of the token definition
+
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or scalar
+        """
+        res = {'type': 'tan',
+               'title': 'TAN Token',
+               'description': 'TAN: TANs printed on a sheet '
+                              'of paper.',
+               'init': {},
+               'config': {},
+               'user':  ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       TANACTION.TANTOKEN_COUNT: {
+                           "type": "int",
+                           "desc": _("The number of OTP values, which are "
+                                     "printed on the paper.")
+                       }
```

Comment:
Hmm, does this policy currently have an effect (I think it has the same action name as the corresponding paper token action)

Reply:
The policy actually does not have the same name, since if a token type specific policy does not start with the tokentype, the tokentype is automatically added as prefix. So this policy results in the action name "tan_papertoken_count". I think nevertheless I should rename it.

But you are right. We are missing a decorator for this policy!

Reply:
done!